src/backend/metal/src/soft.rs

```diff
@@ -1,49 +1,75 @@
 use command::{IndexBuffer};
-use native::{Frame, ImageRoot, RasterizerState};
+use native::{Frame, ImageRoot, ImageRootRef, RasterizerState};
 
 use hal;
 use metal;
 
 use std::ops::Range;
 
-pub fn push_data(constants: &[u32]) -> Vec<u8> {
-    constants
-        .iter()
-        .flat_map(|&v| (0 .. 4).map(move |i| (v >> 8*i) as u8))
-        .collect()
+
+pub trait Resources {
+    type Data;
+    type Buffer;
+    type Texture;
+    type Sampler;
+    type DepthStencil;
+    type RenderPipeline;
+    type ComputePipeline;
 }
 
+#[derive(Debug)]
+pub enum Own {}
+impl Resources for Own {
+    type Data = Vec<u32>;
+    type Buffer = metal::Buffer;
+    type Texture = ImageRoot;
+    type Sampler = metal::SamplerState;
+    type DepthStencil = metal::DepthStencilState;
+    type RenderPipeline = metal::RenderPipelineState;
+    type ComputePipeline = metal::ComputePipelineState;
+}
+impl<'a> Resources for &'a Own {
+    type Data = &'a [u32];
+    type Buffer = &'a metal::BufferRef;
+    type Texture = ImageRootRef<'a>;
+    type Sampler = &'a metal::SamplerStateRef;
+    type DepthStencil = &'a metal::DepthStencilStateRef;
+    type RenderPipeline = &'a metal::RenderPipelineStateRef;
+    type ComputePipeline = &'a metal::ComputePipelineStateRef;
+}
+
+
 #[derive(Clone, Debug)]
-pub enum RenderCommand {
+pub enum RenderCommand<R: Resources> {
     SetViewport(metal::MTLViewport),
     SetScissor(metal::MTLScissorRect),
     SetBlendColor(hal::pso::ColorValue),
     SetDepthBias(hal::pso::DepthBias),
-    SetDepthStencilDesc(metal::DepthStencilState),
+    SetDepthStencilDesc(R::DepthStencil),
     SetStencilReferenceValues(hal::pso::StencilValue, hal::pso::StencilValue),
     BindBuffer {
         stage: hal::pso::Stage,
         index: usize,
-        buffer: Option<metal::Buffer>,
+        buffer: Option<R::Buffer>,
         offset: hal::buffer::Offset,
     },
     BindBufferData {
         stage: hal::pso::Stage,
         index: usize,
-        bytes: Vec<u8>,
+        words: R::Data,
```

Comment:
In general I don't think 4-byte alignment is necessary in most places for iOS and tvOS, so we may want to reconsider eventually, or set a `WORD_SIZE` based on the OS